### PR TITLE
Topic names

### PR DESCRIPTION
**Pull request by Karsten1987**
this should fix the inconsistency with `rclcpp::Node::count_subscribers` and `rclcpp::Node::count_publishers` as well as `get_topic_names_and_types`.

The function truncates the internal topic fqdn in the form of `rt/my_namespace/my_topic` into `/my_namespace/my_topic`

@clalancette Could you please review this and report whether this solves your issue?

### CODE REVIEW

**Comment by clalancette** [comment]
So this isn't currently working for my use case.  Looking at it, the reason is that I am calling count_subscribers with the string "map", but this PR removes all of the prefixes, but not the leading "/", and thus it is comparing "map" to "/map", and failing.  How should we handle that?

**Comment by Karsten1987** [comment]
please rebase to the current head of  rcl and rclcpp as Williams topic name
validation got merged. This will prevent you from having this exact
mismatch.

**Comment by clalancette** [comment]
Ah, OK, will do.

**Comment by clalancette** [comment]
Still the same problem with @wjwwood 's code.

**Comment by clalancette** [comment]
I ended up changing _filter_ros_prefix() to look like this:

```
inline
std::string
_filter_ros_prefix(const std::string & topic_name)
{
  std::list<std::string> prefixes{ros_topic_prefix,
      ros_service_requester_prefix, ros_service_response_prefix};

  for (auto it : prefixes) {
    std::string withslash = it + "/";
    if (topic_name.rfind(withslash, 0) == 0) {
      return topic_name.substr(withslash.length());
    }
  }
  return topic_name;
}
```

This has a few benefits:
* Doesn't repeat the substr code for each string we want to filter
* Only finds the prefix at the beginning of the string, not anywhere in the string
* Removes the slash along with the prefix

The last point there is mostly what makes it work for me.  This code still has the problem that if the user does pass in "/map", it still won't match.  I'm not sure exactly what we want to do about that.  It seems to me that we want to write some kind of "canonicalization" function that takes in a topic, and produces a canonical, fully-qualified version based on the rules on http://design.ros2.org/articles/topic_and_service_names.html, but I'm not sure.

**Comment by wjwwood** [comment]
@Karsten1987 you will still need to patch rclcpp to call `expand_topic_or_service_name` within `count_subscriptions` so that @clalancette doesn't have the issue with `map` vs `/map`.

**Comment by Karsten1987** [comment]
@clalancette I see two potential problems with your proposal:

1.) rfind does only find the last occurrence, that means a valid topic such as `rt/rt/rt` does not get truncated as the last occurrence is not 0.
2.) the list is nice, but it copies the values (nitpick here, but still)

@wjwwood Good point. Will do that straight ahead.

**Comment by clalancette** [comment]
@Karsten1987 Actually, I don't think point 1) is a problem.  The reason is that we are calling rfind() with a pos of 0, which, according to http://www.cplusplus.com/reference/string/string/rfind/, means:

```
Position of the last character in the string to be considered as the beginning of a match
```

Thus, only the very beginning of the string will be considered for matching.  In a test program locally, the `rt/rt/rt` topic does get properly truncated to `rt/rt` when using my method above.

File: rmw_fastrtps_cpp/src/functions.cpp [hunk 1]
```diff
@@ -536,14 +548,21 @@ typedef struct CustomParticipantInfo
...
+_filter_ros_prefix(const std::string & topic_name)
 {
-// static for internal linkage
-static const char * const eprosima_fastrtps_identifier = "rmw_fastrtps_cpp";
-static const char * const ros_topic_prefix = "rt";
-static const char * const ros_service_requester_prefix = "rq";
-static const char * const ros_service_response_prefix = "rr";
+  auto prefixes = {ros_topic_prefix, ros_service_requester_prefix, ros_service_response_prefix};
+  for (auto prefix : prefixes) {
+    if (topic_name.rfind(std::string(prefix) + "/") == 0) {
```
Comment (clalancette) [inline rmw_fastrtps_cpp/src/functions.cpp]: As we discussed on chat, this should be:

`if (topic_name.rfind(std::string(prefix) + "/", 0) == 0) {`

**Comment by clalancette** [comment]
Other than the one bug I pointed out in the code, this looks pretty good to me (and, in combination with the changes on the topic_names branch in rclcpp, works for me).  @wjwwood , I'd like to have your opinion on it before we approve.

**Comment by wjwwood** [comment]
@Karsten1987 when do you think you'll be ready to merge this?

I started this CI to see if it helps with the linux packaging job:

[Build Status](http://ci.ros2.org/job/ci_packaging_linux/3/)

**Comment by clalancette** [comment]
@wjwwood That job ran afoul of the setuptools issue, I kicked it off again here:


[Build Status](http://ci.ros2.org/job/ci_packaging_linux/5/)

**Comment by Karsten1987** [comment]
this one seems to compile:
[Build Status](http://ci.ros2.org/job/ci_packaging_linux/6/)

**Comment by Karsten1987** [comment]
The dynamic bridge seems to constantly timeout. Is that a known flaky test?

**Comment by dhood** [comment]
the tests broke when the /rt prefixes were added

**Comment by Karsten1987** [comment]
Then I think this PR is unrelated. Fixing the dynamic bridge tests should be in a separate PR. Thoughts?

**Comment by clalancette** [comment]
@Karsten1987 Yeah, I agree.  If you look at the previous builds (like http://ci.ros2.org/view/packaging/job/packaging_linux/695/#showFailuresLink), that same test fails.  So this PR doesn't make things worse, I think it should be good to merge.

File: rmw_fastrtps_cpp/src/functions.cpp [hunk 2]
```diff
@@ -1799,6 +1816,7 @@ rmw_client_t * rmw_create_client(const rmw_node_t * node,
 
   info->writer_guid_ = info->request_publisher_->getGuid();
 
+  fprintf(stderr, "Original service lcient topic %s\n", service_name);
```
Comment (wjwwood) [inline rmw_fastrtps_cpp/src/functions.cpp]: These should not be merged in, here and below.

**Comment by Karsten1987** [comment]
[Build Status](http://ci.ros2.org/job/ci_windows/2780/)
[Build Status](http://ci.ros2.org/job/ci_linux/2662/)
[Build Status](http://ci.ros2.org/job/ci_osx/2115/)

**Pull request merged**